lib/srv/db/redis/engine.go

```diff
@@ -0,0 +1,468 @@
+		} else if err != nil {
+			// Send the error to the client as connection will be terminated after return.
+			e.SendError(err)
+
+			return trace.Wrap(err)
+		} else {
+			vals, err = cmd.Result()
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+
+		// Send response back to the client.
+		if err := e.sendToClient(vals); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+// writeCmd writes Redis commands passed as vals to Redis wire form.
+// Most types is covered by go-redis implemented WriteArg() function. Types override by this function are:
+// * Redis errors and Go error: go-redis returns a "human-readable" string instead of RESP compatible error message
+// * integers: go-redis converts them to string, which is not always what we want to.
+// * slices: arrays are recursively converted to RESP responses.
+func writeCmd(wr *redis.Writer, vals interface{}) error {
+	switch val := vals.(type) {
+	case redis.Error:
+		if val == redis.Nil {
+			// go-redis returns nil value as errors, but Redis Wire protocol decodes them differently.
+			// Note: RESP3 has different sequence for nil.
+			if _, err := wr.WriteString("$-1\r\n"); err != nil {
+				return trace.Wrap(err)
+			}
+			return nil
+		}
+
+		if err := wr.WriteByte('-'); err != nil {
+			// Redis error passed from DB itself. Just add '-' to mark as error.
+			return trace.Wrap(err)
+		}
+
+		if _, err := wr.WriteString(val.Error()); err != nil {
+			return trace.Wrap(err)
+		}
+
+		if _, err := wr.Write([]byte("\r\n")); err != nil {
+			return trace.Wrap(err)
+		}
+	case error:
+		if _, err := wr.WriteString("-ERR "); err != nil {
+			// Add error header specified in https://redis.io/topics/protocol#resp-errors
+			// to follow the convention.
+			return trace.Wrap(err)
+		}
+
+		if _, err := wr.WriteString(val.Error()); err != nil {
+			return trace.Wrap(err)
+		}
+
+		if _, err := wr.Write([]byte("\r\n")); err != nil {
+			return trace.Wrap(err)
+		}
+	case int:
+		if err := writeInteger(wr, int64(val)); err != nil {
+			return trace.Wrap(err)
+		}
+	case int8:
+		if err := writeInteger(wr, int64(val)); err != nil {
+			return trace.Wrap(err)
+		}
+	case int16:
+		if err := writeInteger(wr, int64(val)); err != nil {
+			return trace.Wrap(err)
+		}
+	case int32:
+		if err := writeInteger(wr, int64(val)); err != nil {
+			return trace.Wrap(err)
+		}
+	case int64:
+		if err := writeInteger(wr, val); err != nil {
+			return trace.Wrap(err)
+		}
+	case []interface{}:
+		if err := wr.WriteByte(redis.ArrayReply); err != nil {
+			return trace.Wrap(err)
+		}
+		n := len(val)
+		if err := wr.WriteLen(n); err != nil {
+			return trace.Wrap(err)
+		}
+
+		for _, v0 := range val {
+			if err := writeCmd(wr, v0); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	case interface{}:
+		err := wr.WriteArg(val)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// writeInteger converts integers to Redis wire form.
+func writeInteger(wr *redis.Writer, val int64) error {
```

Comment:
Can we isolate all functions that deal with converting to/from Redis wire format in its own `protocol` subpackage?